### PR TITLE
Support Flexible Voting

### PR DESCRIPTION
**Pull request by alexkeating**
#### Description

- Support fractional voting in the L2 vote aggregator

### CODE REVIEW

**Comment by alexkeating** [comment]
Current dependencies on/for this PR:
* `main`
  * **PR #93** <a href="https://app.graphite.dev/github/pr/ScopeLift/l2-flexible-voting/93?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/ScopeLift/l2-flexible-voting/93?utm_source=stack-comment).

File: src/L2CountingFractional.sol [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+    bytes memory voteData
+  ) internal {
+    require(voteData.length == 48, "GovernorCountingFractional: invalid voteData");
+
+    (uint128 _againstVotes, uint128 _forVotes, uint128 _abstainVotes) = _decodePackedVotes(voteData);
+
+    uint128 _existingWeight = _proposalVotersWeightCast[proposalId][account];
+    uint256 _newWeight = uint256(_againstVotes) + _forVotes + _abstainVotes + _existingWeight;
+
+    require(_newWeight <= totalWeight, "GovernorCountingFractional: vote would exceed weight");
```
Comment (wildmolasses) [inline src/L2CountingFractional.sol:178]:
```suggestion
    require(_newWeight <= totalWeight, "L2CountingFractional: vote would exceed weight");
```

File: src/L2CountingFractional.sol [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+   * Note that if partial votes are cast, all remaining weight must be cast
+   * with _countVoteFractional: _countVoteNominal will revert.
+   */
+  function _countVoteFractional(
+    uint256 proposalId,
+    address account,
+    uint128 totalWeight,
+    bytes memory voteData
+  ) internal {
+    require(voteData.length == 48, "GovernorCountingFractional: invalid voteData");
```
Comment (wildmolasses) [inline src/L2CountingFractional.sol:171]:
```suggestion
    require(voteData.length == 48, "L2CountingFractional: invalid voteData");
```

File: src/L2CountingFractional.sol [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+    _proposalVotersWeightCast[proposalId][account] = totalWeight;
+
+    if (support == uint8(GovernorCompatibilityBravo.VoteType.Against)) {
+      _proposalVotes[proposalId].againstVotes += totalWeight;
+    } else if (support == uint8(GovernorCompatibilityBravo.VoteType.For)) {
+      _proposalVotes[proposalId].forVotes += totalWeight;
+    } else if (support == uint8(GovernorCompatibilityBravo.VoteType.Abstain)) {
+      _proposalVotes[proposalId].abstainVotes += totalWeight;
+    } else {
+      revert("GovernorCountingFractional: invalid support value, must be included in VoteType enum");
```
Comment (wildmolasses) [inline src/L2CountingFractional.sol:140]:
```suggestion
      revert("L2CountingFractional: invalid support value, must be included in VoteType enum");
```

File: src/L2CountingFractional.sol [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+   */
+  function _countVoteNominal(
+    uint256 proposalId,
+    address account,
+    uint128 totalWeight,
+    uint8 support
+  ) internal {
+    require(
+      _proposalVotersWeightCast[proposalId][account] == 0,
+      "GovernorCountingFractional: vote would exceed weight"
```
Comment (wildmolasses) [inline src/L2CountingFractional.sol:128]:
```suggestion
      "L2CountingFractional: vote would exceed weight"
```

File: src/L2CountingFractional.sol [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+  function _countVote(
+    uint256 proposalId,
+    address account,
+    uint8 support,
+    uint256 totalWeight,
+    bytes memory voteData
+  ) internal virtual {
+    require(totalWeight > 0, "GovernorCountingFractional: no weight");
+    if (_proposalVotersWeightCast[proposalId][account] >= totalWeight) {
+      revert("GovernorCountingFractional: all weight cast");
```
Comment (wildmolasses) [inline src/L2CountingFractional.sol:104]:
```suggestion
      revert("L2CountingFractional: all weight cast");
```

File: src/L2CountingFractional.sol [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+  // Made both of these internal
+  mapping(uint256 => mapping(address => uint128)) internal _proposalVotersWeightCast;
+
+  /**
+   * @dev Mapping from voter address to signature-based vote nonce. The
+   * voter's nonce increments each time a signature-based vote is cast with
+   * fractional voting params and must be included in the `params` as the last
+   * 16 bytes when signing for a fractional vote.
+   */
+  mapping(address => uint128) public fractionalVoteNonce;
```
Comment (wildmolasses) [inline src/L2CountingFractional.sol:36]: unneeded, can remove

**Review by wildmolasses — CHANGES_REQUESTED** [review]
a few nits otherwise lgtm

File: src/L2VoteAggregator.sol [hunk 7]
```diff
@@ -58,25 +59,31 @@ abstract contract L2VoteAggregator is EIP712, L2GovernorMetadata {
     INVALID_Executed
   }
 
-  /// @dev Data structure to store vote preferences expressed by depositors.
```
[the remark this answers is not in the document]
Comment (alexkeating) [reply src/L2VoteAggregator.sol:61]: Good point, I think we are okay with that for now

**Pull request merged**